bug:

├── tf_sample
│   └── main.tf
└── tfsec_cus
    └── test_tfchecks.json


**Contents of main.tf**

resource "aws_security_group" "test" {
  name        = "allow-all-sg"
  description = "Allowing everyone to connect to this public instance"
  vpc_id      = aws_vpc.my_vpc.id
  ingress {
    cidr_blocks = ["0.0.0.0/0"]
    from_port   = "22"
    to_port     = "22"
    protocol    = "tcp"
  }
  tags = {
    Type = "Public"
  }

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}


**Content of test_tfchecks.json**

  "checks": [
    {
      "code": "CUS001",
      "description": "Custom check to ensure the CostCentre tag is applied to SG",
      "impact": "By not having CostCentre we can't keep track of billing",
      "resolution": "Add the CostCentre tag",
      "requiredTypes": [
        "resource"
      ],
      "requiredLabels": [
        "aws_security_group"
      ],
      "severity": "ERROR",
      "matchSpec": {
        "name": "tags",
        "action": "contains",
        "value": "CostCentre"
      },
      "errorMessage": "The required CostCentre tag was missing",
      "relatedLinks": [
        "http://internal.acmecorp.com/standards/aws/tagging.html"
      ]
    }
  ]   
}

Output:

Result #1 CRITICAL Security group rule allows egress to multiple public internet addresses. 
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 main.tf Line 19
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   15  │   egress {
   16  │     from_port   = 0
   17  │     to_port     = 0
   18  │     protocol    = "-1"
   19  │     cidr_blocks = ["0.0.0.0/0"]
   20  │   }
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
          ID aws-vpc-no-public-egress-sgr
      Impact Your port is egressing data to the internet
  Resolution Set a more restrictive cidr range

  More Information
  - https://aquasecurity.github.io/tfsec/v1.0.2/checks/aws/vpc/no-public-egress-sgr/
  - https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


Result #2 CRITICAL Security group rule allows ingress from public internet. 
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 main.tf Line 6
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
    5  │   ingress {
    6  │     cidr_blocks = ["0.0.0.0/0"]
    7  │     from_port   = "22"
    8  │     to_port     = "22"
    9  │     protocol    = "tcp"
   10  │   }
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
          ID aws-vpc-no-public-ingress-sgr
      Impact Your port exposed to the internet
  Resolution Set a more restrictive cidr range

  More Information
  - https://aquasecurity.github.io/tfsec/v1.0.2/checks/aws/vpc/no-public-ingress-sgr/
  - https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule#cidr_blocks
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


Result #3 HIGH Custom check failed for resource aws_security_group.test. The required CostCentre tag was missing 
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 main.tf Lines 1-21
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
    1  │ resource "aws_security_group" "test" {
    2  │   name        = "allow-all-sg"
    3  │   description = "Allowing everyone to connect to this public instance"
    4  │   vpc_id      = aws_vpc.my_vpc.id
    5  │   ingress {
    6  │     cidr_blocks = ["0.0.0.0/0"]
    7  │     from_port   = "22"
    8  │     to_port     = "22"
    9  │     protocol    = "tcp"
   10  │   }
   11  │   tags = {
   12  │     Type = "Public"
   13  │   }
   14  │ 
   15  │   egress {
   16  │     from_port   = 0
   17  │     to_port     = 0
   18  │     protocol    = "-1"
   19  │     cidr_blocks = ["0.0.0.0/0"]
   20  │   }
   21  │ }
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
          ID custom-custom-cus001
      Impact By not having CostCentre we can't keep track of billing
  Resolution Add the CostCentre tag
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x166e663]

goroutine 1 [running]:
github.com/aquasecurity/tfsec/internal/app/tfsec/cmd.outputFormat.func1({{{0x0, 0x0}, {0xc000402094, 0x6}, {0xc000023f00, 0x3a}, {0x0, 0x0}, {0xc000023f80, 0x37}, ...}, ...})
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/internal/app/tfsec/cmd/output.go:56 +0x3c3
github.com/aquasecurity/defsec/formatters.(*base).GetLinks(_, {{{0x0, 0x0}, {0xc000402094, 0x6}, {0xc000023f00, 0x3a}, {0x0, 0x0}, {0xc000023f80, ...}, ...}, ...})
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/vendor/github.com/aquasecurity/defsec/formatters/formatter.go:64 +0x5a
github.com/aquasecurity/defsec/formatters.printResult({0x1918768, 0xc0003fe340}, {0x3, 0x3, {0xc00051e900, 0x1, 0x1}})
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/vendor/github.com/aquasecurity/defsec/formatters/default.go:166 +0x107d
github.com/aquasecurity/defsec/formatters.outputDefault({0x1918768, 0xc0003fe340}, {0xc0001cb200, 0x5, 0x8})
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/vendor/github.com/aquasecurity/defsec/formatters/default.go:44 +0x674
github.com/aquasecurity/defsec/formatters.(*base).Output(0x0, {0xc0001cb200, 0x0, 0x8})
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/vendor/github.com/aquasecurity/defsec/formatters/formatter.go:75 +0x69
github.com/aquasecurity/tfsec/internal/app/tfsec/cmd.outputFormat(0x0, {0x0, 0x0}, {0x0, 0x10}, {0xc0000251a0, 0x22}, {0xc0001cb200, 0x5, 0x8})
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/internal/app/tfsec/cmd/output.go:104 +0x794
github.com/aquasecurity/tfsec/internal/app/tfsec/cmd.output({0x0, 0x0}, {0xc0000aa840, 0x1, 0xc00035f578}, {0xc0000251a0, 0x22}, {0xc0001cb200, 0x5, 0x8})
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/internal/app/tfsec/cmd/output.go:23 +0x246
github.com/aquasecurity/tfsec/internal/app/tfsec/cmd.glob..func1(0x1e8b200, {0xc000399f20, 0x1, 0x3})
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/internal/app/tfsec/cmd/root.go:156 +0xafd
github.com/spf13/cobra.(*Command).execute(0x1e8b200, {0xc00001e090, 0x3, 0x3})
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/vendor/github.com/spf13/cobra/command.go:856 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0x1e8b200)
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/vendor/github.com/spf13/cobra/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/vendor/github.com/spf13/cobra/command.go:902
main.main()
	/private/tmp/tfsec-20220130-14370-1d4ngkq/tfsec-1.0.2/cmd/tfsec/main.go:11 +0x25

--- a/internal/app/tfsec/cmd/output.go
+++ b/internal/app/tfsec/cmd/output.go
@@ -45,6 +45,10 @@ func outputFormat(addExtension bool, baseFilename string, format string, dir str
 			if version.Version != "" {
 				v = version.Version
 			}
+			var links []string
+			if result.Rule().Terraform != nil {
+				links = result.Rule().Terraform.Links
+			}
 			return append([]string{
 				fmt.Sprintf(
 					"https://aquasecurity.github.io/tfsec/%s/checks/%s/%s/%s/",
@@ -53,7 +57,7 @@ func outputFormat(addExtension bool, baseFilename string, format string, dir str
 					strings.ToLower(result.Rule().Service),
 					result.Rule().ShortCode,
 				),
-			}, result.Rule().Terraform.Links...)
+			}, links...)
 		}).
 		WithBaseDir(dir).
 		WithMetricsEnabled(!conciseOutput)

--- a/internal/app/tfsec/cmd/output.go
+++ b/internal/app/tfsec/cmd/output.go
@@ -34,31 +34,33 @@ func output(baseFilename string, formats []string, dir string, results []rules.R
 	return nil
 }
 
+func gatherLinks(result rules.Result) []string {
+	v := "latest"
+	if version.Version != "" {
+		v = version.Version
+	}
+	var links []string
+	if result.Rule().Terraform != nil {
+		links = result.Rule().Terraform.Links
+	}
+	return append([]string{
+		fmt.Sprintf(
+			"https://aquasecurity.github.io/tfsec/%s/checks/%s/%s/%s/",
+			v,
+			result.Rule().Provider,
+			strings.ToLower(result.Rule().Service),
+			result.Rule().ShortCode,
+		),
+	}, links...)
+}
+
 func outputFormat(addExtension bool, baseFilename string, format string, dir string, results []rules.Result) (string, error) {
 
 	formatter := formatters.New().
 		WithDebugEnabled(debug.Enabled).
 		WithColoursEnabled(!disableColours).
 		WithGroupingEnabled(!disableGrouping).
-		WithLinksFunc(func(result rules.Result) []string {
-			v := "latest"
-			if version.Version != "" {
-				v = version.Version
-			}
-			var links []string
-			if result.Rule().Terraform != nil {
-				links = result.Rule().Terraform.Links
-			}
-			return append([]string{
-				fmt.Sprintf(
-					"https://aquasecurity.github.io/tfsec/%s/checks/%s/%s/%s/",
-					v,
-					result.Rule().Provider,
-					strings.ToLower(result.Rule().Service),
-					result.Rule().ShortCode,
-				),
-			}, links...)
-		}).
+		WithLinksFunc(gatherLinks).
 		WithBaseDir(dir).
 		WithMetricsEnabled(!conciseOutput)
 

